c++: Add connection graph

bryfox opened the pull request:
### Changelog

C++: Add connection graph support to the WS server

### Docs
### Description

This adds connection graph support for C++ and an `example_connection_graph` target. The graph can be rendered in the Topic Graph panel of Foxglove.

eloff commented on cpp/foxglove/include/foxglove/server.hpp:
```diff
@@ -22,6 +23,22 @@ struct ClientChannel {
   size_t schemaLen;
 };
 
+class ConnectionGraph final {
```
Should go in it's own .hpp/.cpp files

eloff commented on c/include/foxglove-c/foxglove-c.h:
```diff
@@ -287,6 +294,63 @@ void foxglove_internal_register_cpp_wrapper(void);
+ */
+foxglove_error foxglove_connection_graph_set_subscribed_topic(struct foxglove_connection_graph *FOXGLOVE_NONNULL graph,
+                                                              struct foxglove_string topic,
+                                                              const struct foxglove_string *subscriber_ids,
```
Should we do this, or should we create a FoxgloveStringArray type?

bryfox replied:
I can go either way on this.

There's no practical safety difference, right? I assume a caller would create them in exactly the same way, so it's a matter of preference for the interface? If we had functions that accepted multiple arrays of strings, I could see a concrete benefit, but I don't see anything like that coming up.

I imagine passing `size_t` args next to arrays is reasonably common — moreso than with the sized string interfaces which led to foxglove_string. So this doesn't seem too bad to me; and we won't end up with different array types for each underlying type.

I guess the other consideration is what we want the callback interface to look like; e.g., how you receive a list of parameter names.

Do you have a preference?

eloff reviewed (commented):
This is looking good so far

gasmith commented on c/src/connection_graph.rs:
```diff
@@ -0,0 +1,271 @@
+    }
+}
+
+/// Create a new `FoxgloveConnectionGraph`.
```
Ultra-pedantry: For docstrings on public C API functions, if we want to refer to type names directly, we should use the mangled snake_case form. In this particular case, I think I'd prefer to just say "Creates a new connection graph" and let the type signature speak for itself.

gasmith commented on c/src/connection_graph.rs:
```diff
@@ -0,0 +1,271 @@
+    let graph_box = Box::new(FoxgloveConnectionGraph::default());
+    unsafe { *graph = Box::into_raw(graph_box) };
+
+    if graph.is_null() {
+        return FoxgloveError::Unspecified;
+    }
```
Is this actually a concern? Rust just panics if it fails to allocate. Maybe in the future we'll have [`Box::try_new`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.try_new), but I think for now it makes sense to just unconditionally return `FoxgloveError::Ok`.

gasmith replied:
Ach, misread. In that case, it needs to happen before we dereference. ;)

bryfox replied:
Ah, thanks; I misunderstood the earlier feedback. I fixed the ordering and kept the error return value since that's the pattern we've adopted elsewhere.

eloff replied:
Btw, ValueError is what we used elsewhere, indicating an issue with a parameter

bryfox replied:
👍 makes sense

gasmith commented on c/src/connection_graph.rs:
```diff
@@ -0,0 +1,271 @@
+/// Overwrites any existing topic with the same name.
+///
+/// # Safety
+/// `topic`, and each ID in `publisher_ids` must adhere to the safety rules of `FoxgloveString`.
```
We should mention that we copy the strings out of these pointers, and so the pointers only need to be valid for the duration of this function call.

gasmith commented on c/src/lib.rs:
```diff
@@ -287,6 +300,29 @@ pub extern "C" fn foxglove_server_stop(
 
+#[unsafe(no_mangle)]
+pub extern "C" fn foxglove_server_publish_connection_graph(
+    server: Option<&mut FoxgloveWebSocketServer>,
+    graph: Option<&mut FoxgloveConnectionGraph>,
+) -> FoxgloveError {
+    let Some(server) = server else {
+        tracing::error!("foxglove_server_publish_connection_graph called with null server");
+        return FoxgloveError::ValueError;
+    };
+    let Some(graph) = graph else {
+        tracing::error!("foxglove_server_publish_connection_graph called with null graph");
+        return FoxgloveError::ValueError;
+    };
```
What's the idea behind accepting an `Option<&mut T>` instead of just `&mut T`?

eloff replied:
The pattern is used all over in the C API, it converts a null pointer to None so you can check for that.

gasmith commented on c/src/lib.rs:
```diff
@@ -287,6 +300,29 @@ pub extern "C" fn foxglove_server_stop(
+        tracing::error!("foxglove_server_publish_connection_graph called with closed server");
+        return FoxgloveError::SinkClosed;
+    };
+    match server.publish_connection_graph(graph.0.clone()) {
```
Could we consume the graph instead of cloning it, so that the caller doesn't need to call `foxglove_connection_graph_free`?

bryfox replied:
Maybe? Right now (in C++) the unique_ptr deals with the freeing so you don't need to think about it. But you also don't need to allocate a new connection graph if you want to re-publish it periodically (see example).

eloff commented on cpp/foxglove/include/foxglove/server/connection_graph.hpp:
```diff
@@ -0,0 +1,33 @@
+public:
+  ConnectionGraph();
+
+  foxglove_connection_graph& impl();
```
Yeah, C++ doesn't have Rust's pub(crate) which is so useful. I vote we allow friend for that purpose.

I never subscribed to the puritan style of OOO in C++ and Java where each class is a strict boundary layer.

bryfox replied:
OK, I've declared WebSocketServer a friend of ConnectionGraph. I think that makes sense here; the latter is really only meaningful in the context of the server.

eloff commented on c/src/connection_graph.rs:
```diff
@@ -19,11 +19,12 @@ pub unsafe extern "C" fn foxglove_connection_graph_create(
-    unsafe { *graph = Box::into_raw(graph_box) };
-
     if graph.is_null() {
         return FoxgloveError::Unspecified;
```
Use ValueError like the other cases, it indicates an issue with the params

eloff reviewed (approved):
This looks good to me, I just left a comment about fixing the error code for the null check

bryfox commented:
I think I've addressed all feedback now, and per https://github.com/foxglove/foxglove-sdk/pull/379#discussion_r2054665633 replaced the impl getter with a friend declaration.
